llvm/lib/SYCL/KernelPropGen/KernelPropGen.cpp

```diff
@@ -0,0 +1,147 @@
+//===- KernelPropGen.cpp                                      ---------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// Retrieves the names of the kernels inside of the passed in file and places
+// them into a text file. Possible to merge this into another pass if
+// required, as it's a fairly trivial pass on its own.
+// ===---------------------------------------------------------------------===//
+
+#include <cstddef>
+#include <regex>
+#include <string>
+
+#include "llvm/SYCL/KernelPropGen.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+// Put the code in an anonymous namespace to avoid polluting the global
+// namespace
+namespace {
+
+/// Retrieve the names for all kernels in the module and place them into a file
+struct KernelPropGen : public ModulePass {
+
+  static char ID; // Pass identification, replacement for typeid
+
+  KernelPropGen() : ModulePass(ID) {}
+
+  bool isKernel(const Function &F) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL)
+      return true;
+    return false;
+  }
+
+  int GetWriteStreamID(StringRef Path) {
+    int FileFD = 0;
+    std::error_code EC =
+          llvm::sys::fs::openFileForWrite(Path, FileFD);
+    if (EC) {
+      llvm::errs() << "Error in KernelPropGen Pass: " << EC.message() << "\n";
+    }
+
+    return FileFD;
+  }
+
+  /// Visit all the functions of the module
+  bool runOnModule(Module &M) override {
+    SmallString<256> TDir;
+    llvm::sys::path::system_temp_directory(true, TDir);
+    std::string file = "KernelProperties_" + M.getSourceFileName();
+    llvm::sys::path::append(TDir, file);
+    llvm::sys::path::replace_extension(TDir, "sh",
+      llvm::sys::path::Style::native);
+
+    llvm::raw_fd_ostream O(GetWriteStreamID(TDir.str()),
+                            true /*close in destructor*/);
+
+    if (O.has_error())
+      return false;
+  
+    // Generate a bash script to inject environment information containing 
+    // kernel properties that we need so we can compile. This script is called
+    // from sycl-xocc. #!/bin/bash
+    O << "#!/bin/bash\n";
+   
+    std::string kernelNames = "";
```

Comment:
Perhaps using `llvm::Twine` around is more in phase with LLVM coding rules.

Reply:
This gave me a reason to read the manual section on llvm's string containers, something I had to do at some point.. I found them quite confusing when using them in the initial driver implementation as they don't quite interface with each other perfectly: http://llvm.org/docs/ProgrammersManual.html#string-like-containers 

Seems like both llvm::Twine and StringRef are not too useful in this case, llvm::Twine is for inline concatenation (and has no move or assignment operators, from what I can tell the assignment operator is deleted and the move isn't implemented and the compiler won't generate a default because of the guidelines on move operators) and StringRef is a reference to a constant string with no ownership so not really built for appending etc.

SmallString seems fine though, you are correct it does indeed function like a vector! Sadly though it doesn't interface particularly well with llvm::Twine (I wonder why though as it works well with StringRef) and the RHS of the += operators are llvm::Twine. So in the end the slightly nasty syntax of having to use .str() remains the same! It should be more efficient though and it's more consistent as it's not reliant on the host's std library.

It is fine to use std::string in LLVM though it seems, when you wish to maintain persistent heap memory, but in this case we don't really care about that.